pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,215 @@
+// Provide returns a credentialprovider.DockerConfig based on the credentials returned
+// by the exec plugin.
+func (p *pluginProvider) Provide(image string) credentialprovider.DockerConfig {
+	cfg := make(credentialprovider.DockerConfig)
+
+	if matched, _ := credentialprovider.URLsMatchStr(p.matchURL, image); !matched {
+		return cfg
+	}
+
+	response, err := p.plugin.ExecPlugin(image)
+	if err != nil {
+		klog.Errorf("Failed getting credential from external registry credential provider: %v", err)
+		return cfg
+	}
+
+	cfg[p.matchURL] = credentialprovider.DockerConfigEntry{
+		Username: response.Username,
+		Password: response.Password,
+		Email:    "plugin@criauth.k8s.io", // email is deprecated, this is just a dummy email
+	}
+	return cfg
+}
+
+// Enabled always returns true since registration of the plugin via kubelet implies it should be enabled.
+func (e *pluginProvider) Enabled() bool {
+	return true
+}
+
+// IsPlugin returns true since all external credential providers are plugins
+func (e *pluginProvider) IsPlugin() bool {
+	return true
+}
+
+// Plugin is the interface calling ExecPlugin. This is mainly for testability
+// so tests don't have to actually exec any processes.
+type Plugin interface {
+	ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error)
+}
+
+// execPlugin is the implementation of the Plugin interface that execs a provided CRI auth plugin based
+// on it's name provided in CRIAuthPluginConfig. It is assumed that the executable is available in the
+// plugin directory provided by the kubelet.
+type execPlugin struct {
+	name         string
+	args         []string
+	pluginBinDir string
+}
+
+// ExecPlugin executes the plugin binary in the following format:
+//
+//  $ <plugin-name> get-credentials arg1=value1 arg2=value2
+//
+// The plugin is expected to receive the CRIAuthPluginRequest API via stdin and is expected
+// to return the CRIAuthPluginResponse via stdout.
+func (e *execPlugin) ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error) {
+	pluginBin := e.pluginBinDir + "/" + e.name
+	args := []string{"get-credentials"}
+	args = append(args, e.args...)
+
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	stdin := &bytes.Buffer{}
+	cmd := exec.Command(pluginBin, args...)
+	cmd.Stdout, cmd.Stderr, cmd.Stdin = stdout, stderr, stdin
+
+	authRequest := &criauth.CRIAuthPluginRequest{Image: image}
+	data, err := encodeRequest(authRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode auth request: %v", err)
+	}
+
+	if _, err := stdin.Write(data); err != nil {
+		return nil, fmt.Errorf("error writing to stdin: %v", err)
+	}
+
+	if err = cmd.Run(); err != nil {
+		klog.V(2).Infof("Error execing CRI auth plugin, stderr: %v", stderr.String())
+		return nil, fmt.Errorf("error execing CRI auth plugin %s for image %s: %v", e.name, image, err)
+	}
+
+	response, err := decodeResponse(stdout.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("error decoding request from stdout: %v", err)
```

Comment:
let's not do it. This will expose the response to the log file. Response will contain credentials

Reply:
I mean not include the `err` to the error message

Reply:
Maybe even add an explicit test that validates that no part of bad formed response was logged

Reply:
Updated to not include err